src/main/java/io/deephaven/csv/sinks/ArraySinkFactory.java

```diff
@@ -0,0 +1,458 @@
+        final ArrayLongSink result = new ArrayLongSink(longSentinel);
+        source.setValue(result);
+        return result;
+    }
+
+    @Override
+    public Long reservedLong() {
+        return longSentinel;
+    }
+
+    @Override
+    public Sink<float[]> forFloat() {
+        return new ArrayFloatSink(floatSentinel);
+    }
+
+    @Override
+    public Float reservedFloat() {
+        return floatSentinel;
+    }
+
+    @Override
+    public Sink<double[]> forDouble() {
+        return new ArrayDoubleSink(doubleSentinel);
+    }
+
+    @Override
+    public Double reservedDouble() {
+        return doubleSentinel;
+    }
+
+    @Override
+    public Sink<byte[]> forBooleanAsByte() {
+        return new ArrayBooleanAsByteSink(booleanAsByteSentinel);
+    }
+
+    @Override
+    public Sink<char[]> forChar() {
+        return new ArrayCharSink(charSentinel);
+    }
+
+    @Override
+    public Character reservedChar() {
+        return charSentinel;
+    }
+
+    @Override
+    public Sink<String[]> forString() {
+        return new ArrayStringSink(stringSentinel);
+    }
+
+    @Override
+    public String reservedString() {
+        return stringSentinel;
+    }
+
+    @Override
+    public Sink<long[]> forDateTimeAsLong() {
+        return new ArrayDateTimeAsLongSink(dateTimeAsLongSentinel);
+    }
+
+    @Override
+    public Long reservedDateTimeAsLong() {
+        return dateTimeAsLongSentinel;
+    }
+
+    @Override
+    public Sink<long[]> forTimestampAsLong() {
+        return new ArrayTimestampAsLongSink(timestampAsLongSentinel);
+    }
+
+    @Override
+    public Long reservedTimestampAsLong() {
+        return timestampAsLongSentinel;
+    }
+}
+
+
+abstract class ArraySinkBase<TARRAY> implements Sink<TARRAY> {
+    private final int INITIAL_SIZE = 1024;
+
+    private final IntFunction<TARRAY> arrayFactory;
+    protected final boolean hasNullSentinel;
+    protected TARRAY array;
+
+    protected ArraySinkBase(final IntFunction<TARRAY> arrayFactory, final boolean hasNullSentinel) {
+        this.arrayFactory = arrayFactory;
+        this.hasNullSentinel = hasNullSentinel;
+        this.array = arrayFactory.apply(INITIAL_SIZE);
+    }
+
+    @Override
+    public final void write(
+            final TARRAY src,
+            final boolean[] isNull,
+            final long destBegin,
+            final long destEnd,
+            boolean appending) {
+        if (destBegin == destEnd) {
+            return;
+        }
+        final int destBeginAsInt = Math.toIntExact(destBegin);
+        final int destEndAsInt = Math.toIntExact(destEnd);
+        final int destSize = Math.toIntExact(destEnd - destBegin);
+
+        final int currentCapacity = Array.getLength(array);
+
+        // Grow array if needed.
+        if (currentCapacity < destEnd) {
+            final int highBit = Integer.highestOneBit(destEndAsInt);
+            final int newCapacity = destEndAsInt == highBit ? destEndAsInt : destEndAsInt * 2;
```

Comment:
This logic feels a bit weird to me. We are essentially saying "if destEndAsInt is a power of two, use that size; otherwise, use 2 * destEndAsInt".

The more canonical way to check power-of-2 is `x & (x - 1) == 0`.

I'd be more inclined to allocate to the power-of-2 that fits destEndAsInt (as opposed to 2x destEndAsInt).

Reply:
oops, nice catch. Fixed.
I was trying to express "round up to the next power of two that is >= destEndInt".
But I didn't want to double it if destEndInt was *already* a power of 2.
So the expression is now
```
            final int highBit = Integer.highestOneBit(destEndAsInt);
            final int newCapacity = destEndAsInt == highBit ? highBit : highBit * 2;
```
I'm aware of the bitmask hack but I don't think that expresses the above any more compactly. Do you have suggested alternate code? Right now I'm going with

Reply:
Maybe not as obvious, but that's what comments are for :)

```java
final int newCapacity = 1 << (32 - Integer.numberOfLeadingZeros(destEndAsInt - 1));
```

I'm happy w/ your version if this is too magic.

Reply:
If it were in an inner loop, and there was evidence that it made the code go faster, I would go with your version.
I think otherwise I would prefer to go with my version because it's a little easier to understand?